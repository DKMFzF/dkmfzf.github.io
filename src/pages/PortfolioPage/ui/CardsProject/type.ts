export interface Card {
	id: number;
	title: string;
	content: string;
	link: string;
	x: number;
	y: number;
	rotation: number;
}

export interface CardsProjectProps {
	cards: Card[];
	transformStyle: string;
}
