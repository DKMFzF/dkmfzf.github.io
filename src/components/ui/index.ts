export * from './preloader';
export * from './modal-overlay';
export * from './pages';
export * from './app-header';
export * from './modal';
export * from './icon';
export * from './link';
export * from './section-about-me';
