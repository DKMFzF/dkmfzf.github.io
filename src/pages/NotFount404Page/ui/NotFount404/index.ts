export { NotFound404Page } from './NotFount404';
