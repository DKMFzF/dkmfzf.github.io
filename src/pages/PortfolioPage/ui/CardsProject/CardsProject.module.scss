@use '~styles/variables' as *;

$premary-card-color: #cec2e6;

.cards-wrapper {
	left: 0;
	position: absolute;
	top: 0;
	transform-origin: center center;
}

.card {
	aspect-ratio: 16 / 9;
	background-color: rgb(28 28 33 / 100%);
	border-radius: $section-border-radius;
	color: white;
	cursor: pointer;
	min-inline-size: 500px;
	overflow: hidden;
	position: absolute;
	text-decoration: none;
	transform-origin: center center;

	&__logo-wrapper {
		block-size: 100%;
		inline-size: 100%;
		position: relative;

		&::before {
			aspect-ratio: 1 / 1;
			background-image: url('../../../../shared/assets/images/stellarburger.png');
			background-position: center;
			background-repeat: no-repeat;
			background-size: contain;
			content: '';
			inline-size: 40%;
			left: 50%;
			position: absolute;
			top: 50%;
			transform: translate(-50%, -50%);
		}
	}

	&__content-wrapper {
		align-items: center;
		background-color: rgb(0 0 0 / 100%);
		block-size: 100%;
		border-radius: 8px;
		box-sizing: border-box;
		display: flex;
		inline-size: 100%;
		left: 0;
		mix-blend-mode: darken;
		opacity: 1;
		padding: 20px;
		pointer-events: none;
		position: absolute;
		top: 0;
		transform: translateX(-100%);
	}

	&__title-container {
		font-family: $font-family;
		font-size: 50px;
	}
}
