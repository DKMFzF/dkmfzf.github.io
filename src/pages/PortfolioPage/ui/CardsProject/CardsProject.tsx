import { useRef, FC } from 'react';
import gsap from 'gsap';
import { CardsProjectProps } from './type';
import styles from './CardsProject.module.scss';

export const CardsProject: FC<CardsProjectProps> = ({
	cards,
	transformStyle
}) => {
	const cardRefs = useRef<Map<number, HTMLAnchorElement>>(new Map());
	const contentWrapper = useRef<Map<number, HTMLDivElement>>(new Map());

	return (
		<ul
			className={styles['cards-wrapper']}
			style={{ transform: transformStyle }}
		>
			{cards.map((card) => (
				<li key={card.id} className={styles.cards__element}>
					<a
						target='_blank'
						ref={(el) => el && cardRefs.current.set(card.id, el)}
						href={card.link}
						className={styles.card}
						style={{
							top: `${card.y}px`,
							left: `${card.x}px`,
							transform: `translate(-50%, -50%) rotate(${card.rotation}deg)`
						}}
						onMouseEnter={() => {
							const el = cardRefs.current.get(card.id);
							const contentContainer = contentWrapper.current.get(
								card.id
							);
							if (el) {
								gsap.to(el, {
									scale: 1.07,
									boxShadow: '0 0 25px rgba(0,0,0,0.2)',
									duration: 0.3,
									ease: 'power3.out'
								});
								el.style.zIndex = '2';
							}

							if (contentContainer) {
								gsap.to(contentContainer, {
									x: 0,
									duration: 0.5,
									ease: 'power3.out'
								});
							}
						}}
						onMouseLeave={() => {
							const el = cardRefs.current.get(card.id);
							const contentContainer = contentWrapper.current.get(
								card.id
							);
							if (el) {
								gsap.to(el, {
									scale: 1,
									boxShadow: '0 4px 10px rgba(0,0,0,0.1)',
									duration: 0.3,
									ease: 'power1.out'
								});
								el.style.zIndex = '1';
							}

							if (contentContainer) {
								gsap.to(contentContainer, {
									x: '-100%',
									duration: 0.5,
									ease: 'power3.in'
								});
							}
						}}
						onMouseDown={(e) => e.stopPropagation()}
					>
						<div className={styles['card__logo-wrapper']}>
							<div
								ref={(el) =>
									el &&
									contentWrapper.current.set(card.id, el)
								}
								className={styles['card__content-wrapper']}
							>
								<div
									className={styles['card__title-container']}
								>
									<h3 className={styles.card__title}>
										{card.title}
									</h3>
								</div>
							</div>
						</div>
					</a>
				</li>
			))}
		</ul>
	);
};
