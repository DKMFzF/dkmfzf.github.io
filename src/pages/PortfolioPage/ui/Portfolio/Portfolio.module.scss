@use '~styles/variables' as *;

.portfolio-content {
	background-color: $bg-color;
	height: 100vh;
	left: 0;
	overflow: hidden;
	position: fixed;
	position: relative;
	top: 0;
	width: 100vw;
}

.heading {
	font-size: 24px;
	font-weight: bold;
	margin: 20px;
}
