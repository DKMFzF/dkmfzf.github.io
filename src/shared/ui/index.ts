export * from './CommonPage';
export * from './Icon';
export * from './Preloader';
// export * from './Text';
