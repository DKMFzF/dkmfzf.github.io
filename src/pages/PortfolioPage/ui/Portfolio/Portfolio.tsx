import { Helmet } from 'react-helmet-async';
import { FC, useState, useEffect } from 'react';
import { CardsProject } from '../CardsProject';
import { portfolioCard } from '../../constants/portfolioCard';
import { META_SITE_DATA } from '@config';
import { useCards, useDragZoom } from '../../model';
import styles from './portfolio.module.scss';

export const PortfolioPage: FC = () => {
	const [isMobileView, setIsMobileView] = useState(false);
	const { isDragging, position, scale, handleMouseDown, containerRef } =
		useDragZoom();
	const { cards: positionedCards } = useCards();

	useEffect(() => {
		const handleResize = () => {
			setIsMobileView(window.innerWidth < 1440);
		};
		handleResize();
		window.addEventListener('resize', handleResize);
		return () => window.removeEventListener('resize', handleResize);
	}, []);

	const cardsWithPosition = portfolioCard.map((originalCard, index) => {
		const positionData =
			positionedCards[Math.min(index, positionedCards.length - 1)];

		return {
			...originalCard,
			id: index,
			x: positionData.x,
			y: positionData.y,
			rotation: positionData.rotation,
			width: positionData.width,
			height: positionData.height
		};
	});

	if (isMobileView) {
		return (
			<>
				<Helmet>
					<title>
						{
							META_SITE_DATA.titles.portfolio
								.portfolioWithNotApproval
						}
					</title>
				</Helmet>
				<main className={styles['mobile-message']} data-cy='content'>
					Please open this page from your PC
				</main>
			</>
		);
	}

	return (
		<>
			<Helmet>
				<title>
					{META_SITE_DATA.titles.portfolio.portfolioWithApproval}
				</title>
			</Helmet>
			<main
				ref={containerRef}
				className={styles.container}
				style={{ cursor: isDragging ? 'grabbing' : 'grab' }}
				onMouseDown={handleMouseDown}
			>
				<CardsProject
					cards={cardsWithPosition}
					transformStyle={`translate(${position.x}px, ${position.y}px) scale(${scale})`}
				/>
			</main>
		</>
	);
};

export default PortfolioPage;
