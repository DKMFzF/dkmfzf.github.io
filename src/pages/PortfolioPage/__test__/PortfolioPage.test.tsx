import { CardsProject } from '../ui/CardsProject';
import { HelmetProvider } from 'react-helmet-async';
import { PortfolioPage } from '../ui/Portfolio/Portfolio';
import { render } from '@testing-library/react';
import { expect } from '@jest/globals';

describe('[CardsProject]', () => {
	it('Renders card-project correctly and matches snapshot', () => {
		const { asFragment } = render(
			<CardsProject
				cards={[
					{
						id: 1,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 110,
						y: 110,
						rotation: 10
					},
					{
						id: 2,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 220,
						y: 220,
						rotation: 20
					},
					{
						id: 3,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 330,
						y: 330,
						rotation: -10
					},
					{
						id: 4,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 440,
						y: 440,
						rotation: -20
					}
				]}
				transformStyle='test-styles'
			/>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});

jest.mock('../model', () => ({
	__esModule: true,
	useDragZoom: jest.fn(() => ({
		isDragging: false,
		position: { x: 0, y: 0 },
		scale: 1,
		handleMouseDown: jest.fn(),
		containerRef: { current: null }
	})),
	useCards: jest.fn(() => ({
		cards: [
			{ x: 0, y: 0, rotation: 0, width: 100, height: 100 },
			{ x: 100, y: 100, rotation: 10, width: 100, height: 100 }
		]
	}))
}));

jest.mock('../ui/CardsProject', () => ({
	__esModule: true,
	CardsProject: jest.fn(() => <div>CardsProject Mock</div>)
}));

describe('[PortfolioPage]', () => {
	beforeEach(() => {
		Object.defineProperty(window, 'innerWidth', {
			writable: true,
			configurable: true,
			value: 1920
		});
	});

	it('should render desktop view correctly', () => {
		const { asFragment } = render(
			<HelmetProvider>
				<PortfolioPage />
			</HelmetProvider>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should render mobile view correctly', () => {
		Object.defineProperty(window, 'innerWidth', {
			writable: true,
			configurable: true,
			value: 800
		});
		const { asFragment } = render(
			<HelmetProvider>
				<PortfolioPage />
			</HelmetProvider>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
