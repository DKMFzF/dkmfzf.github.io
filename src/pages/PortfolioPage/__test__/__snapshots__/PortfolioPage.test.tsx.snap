// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[CardsProject] Renders card-project correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    CardsProject Mock
  </div>
</DocumentFragment>
`;

exports[`[PortfolioPage] should render desktop view correctly 1`] = `
<DocumentFragment>
  <main
    class="container"
    style="cursor: grab;"
  >
    <div>
      CardsProject Mock
    </div>
  </main>
</DocumentFragment>
`;

exports[`[PortfolioPage] should render mobile view correctly 1`] = `
<DocumentFragment>
  <main
    class="mobile-message"
    data-cy="content"
    style=""
  >
    Please open this page from your PC
  </main>
</DocumentFragment>
`;
